clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
+        // We'll fill in the stop ordinal in our second pass below.
+        uint64_t start = entry.instr_count;
+        uint64_t timestamp = entry.timestamp;
+        // Some entries have no instructions.
+        if (all_sched[cur_output].size() > 0 &&
+            input == all_sched[cur_output].back().key.input &&
+            start == all_sched[cur_output].back().start_instruction) {
+            VPRINT(this, 3,
+                   "Output #%d: as-read segment #%zu has no instructions: skipping\n",
+                   cur_output, all_sched[cur_output].size() - 1);
+            continue;
+        }
+        all_sched[cur_output].emplace_back(schedule_record_t::DEFAULT, input, start, 0,
+                                           timestamp);
+        start2stop[input][entry.instr_count] = entry.instr_count;
+    }
+    for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
+        VPRINT(this, 1, "Read %zu as-traced records for output #%d\n",
+               all_sched[i].size(), i);
+        // Update the stop_instruction field and collapse consecutive entries while
+        // inserting into the final location.
+        int start_consec = -1;
+        for (int j = 0; j < static_cast<int>(all_sched[i].size()); ++j) {
+            auto &segment = all_sched[i][j];
+            auto find = start2stop[segment.key.input].find(segment.start_instruction);
+            ++find;
+            if (find == start2stop[segment.key.input].end())
+                segment.stop_instruction = std::numeric_limits<uint64_t>::max();
+            else
+                segment.stop_instruction = find->second;
+            VPRINT(this, 3,
+                   "as-read segment #%d: input=%d start=%" PRId64 " stop=%" PRId64
+                   " time=%" PRId64 "\n",
+                   j, segment.key.input, segment.start_instruction,
+                   segment.stop_instruction, segment.timestamp);
+            if (j + 1 < static_cast<int>(all_sched[i].size()) &&
+                segment.key.input == all_sched[i][j + 1].key.input &&
+                segment.stop_instruction == all_sched[i][j + 1].start_instruction) {
+                // Collapse into next.
+                if (start_consec == -1)
+                    start_consec = j;
+            } else {
+                schedule_record_t &toadd =
+                    start_consec >= 0 ? all_sched[i][start_consec] : all_sched[i][j];
+                outputs_[i].record.emplace_back(
+                    static_cast<typename schedule_record_t::record_type_t>(toadd.type),
+                    +toadd.key.input, +toadd.start_instruction,
+                    +all_sched[i][j].stop_instruction, +toadd.timestamp);
+                start_consec = -1;
+                VDO(this, 3, {
+                    auto &added = outputs_[i].record.back();
+                    VPRINT(this, 3,
+                           "segment #%zu: input=%d start=%" PRId64 " stop=%" PRId64
```

Comment:
Maybe add output-idx to log too?

Also, if one want sto see the final entries but not the as-read entries, use a higher verbosity level above or a lower here.

Reply:
Each output is identified by the "Read %zu..." above.  Raised verbosity above.